### DISCUSSION TITLE
Naming of Generic Types

### DISCUSSION THREAD

**Discussion by MikeNeilson** [opendcs/opendcs#776] [Development Practices]
So sonar is flagging the use of "not a single character" generic type. https://sonarcloud.io/organizations/opendcs/rules?open=java%3AS119&rule_key=java%3AS119 (anyone in the order should be able to get to that, but the basic rule is to enforce this regex on Generic type names: `^[A-Z][0-9]?$`

Personally, I don't think enforcing that is good. Especially coming from the C++ work where template names are common and provide a lot of context in the name it's preferable, except in really obvious cases to treat it like what it is, an identifier of meaning to the reader.

I've done this throughout my improvements to OpenDCS. I've also yet to see an argument beyond "That's the Oracle recommendation" about why to stick with that.

I like what this guy has to say: https://www.vojtechruzicka.com/break-java-generics-naming-convention/
And then there's this: https://stackoverflow.com/questions/40908389/what-is-the-reason-for-using-single-character-generic-type-names-in-java which links to https://en.wikipedia.org/wiki/Naming_convention_(programming)#Readability

I'm not going to arbitrarily change it now, but let the discussion begin. I am inclined to do it if someone can't provide a compelling reason not to. Noting that I don't buy "Oracle Said so", as-is, as a compelling argument.

@opendcs/opendcs-core-devs @opendcs/opendcs-triage

**Comment by adamkorynta** [comment]
I personally don't have a problem with the Oracle recommendation and would like to add on to the reasoning for short letters is because the java frameworks that use generics to denote "don't care" about what type you're dealing with in the internals and externals will be dealing with concrete typing. For this reason, I disagree with Option 3 on https://www.vojtechruzicka.com/break-java-generics-naming-convention/ and I don't think colors alone are enough to distinguish classes from generics (admittedly I'm probably too blue pilled). This falls into the category with ternary expressions and in-line if statements where you are conveying to readers that nothing particularly interesting or complicated is going on here to grok the context.

That said, I put in my hat for the Option 2 outlined in that link allowing a prefix/suffix in the regex, especially when we can start using `var` and save some line space and matches conventions in other languages. 

For Option 3 we'd likely be disabling the check altogether (or matching the regex check for class names) and I think it gives too much weight to generics no matter my IDE's/GitHub's/text editor's color choice.

**Reply by MikeNeilson** [reply]
Yeah, i wouldn't bother with the color thing, but I'm also just used to type names being well, a type on their own.

Do do like the idea of `<name>Type` or just `<name>T` as it provides visual clarity and gives something to enforce. Personally I prefer Type over just T as mentally I do better with things that can read aloud in a normalish sense.

**Reply by adamkorynta** [reply]
I'd pick `T<name>` to match C# (from minimal googling, C++ looks like conventions all over the place)

**Reply by MikeNeilson** [reply]
ugh, I despise such notation, always have. Not that I have a logical reason to. For C++ i would focus on the Standard template library, but even then there a few decades of history.

Let's wait for feedback from others before committing to anything specific.
